compiler/transformation/transform.rs

```diff
@@ -61,3 +62,60 @@ fn prune_redundant_roleplayer_deduplication(conjunction: &mut Conjunction, block
         }
     });
 }
+
+pub fn optimize_away_statically_unsatisfiable_conjunctions(
+    conjunction: &mut Conjunction,
+    block_annotations: &TypeAnnotations,
+) {
+    optimize_away_statically_unsatisfiable_conjunctions_impl(conjunction, block_annotations);
+}
+
+fn optimize_away_statically_unsatisfiable_conjunctions_impl(
+    conjunction: &mut Conjunction,
+    block_annotations: &TypeAnnotations,
+) -> bool {
+    let mut must_optimise_away = false;
+    for nested in conjunction.nested_patterns_mut() {
+        match nested {
+            NestedPattern::Disjunction(disjunction) => {
+                let mut optimised_unsatisfiable_branch_ids = Vec::new();
+                for branch in disjunction.conjunctions_mut().iter_mut() {
+                    if optimize_away_statically_unsatisfiable_conjunctions_impl(branch, block_annotations) {
+                        optimised_unsatisfiable_branch_ids.push(branch.scope_id())
+                    }
+                }
+                disjunction.optimise_away_failing_branches(optimised_unsatisfiable_branch_ids);
+                must_optimise_away = must_optimise_away || disjunction.conjunctions().is_empty();
+            }
+            NestedPattern::Negation(negation) => {
+                optimize_away_statically_unsatisfiable_conjunctions_impl(negation.conjunction_mut(), block_annotations);
+            }
+            NestedPattern::Optional(optional) => {
+                optimize_away_statically_unsatisfiable_conjunctions_impl(optional.conjunction_mut(), block_annotations);
+            }
+        }
+    }
+    let must_optimise_away = must_optimise_away
+        || conjunction.constraints().iter().any(|constraint| {
+            if let Some(constraint_annotation) = block_annotations.constraint_annotations_of(constraint.clone()) {
```

Comment:
If constraints are `Copy`, we should do this:
```suggestion
        || conjunction.constraints().iter().any(|&constraint| {
            if let Some(constraint_annotation) = block_annotations.constraint_annotations_of(constraint) {
```
If they are not, I question why `constraint_annotations_of` does not accept a reference.

Reply:
They're not copy. 
I question that and have a TODO comment on top that I need to get to in a different PR.